### PR TITLE
C++: Further performance improvement for the null termination queries

### PR DESCRIPTION
**Pull request by geoffw0**
Follow-up to https://github.com/github/codeql/pull/6915.

The goal here was to eliminate the mutual recursion between `mayAddNullTerminator` and `mayAddNullTerminatorHelper`, which was still causing unnecessarily slow performance on a few projects.  I'm not sure if I've found the best solution to this - in particular `mayAddNullTerminatorHelper` looks quite large to me, though I haven't seen it blow up in practice.  Happy to take further suggestions.

The change has a *small* positive effect on performance locally.  I ran [DCA](https://github.com/github/codeql-dca-main/issues/969) on this (actually a variant with the two queries enabled) and only saw a 29s overall improvement there, which is likely not statistically significant.

@MathiasVP

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/commons/NullTermination.qll [hunk 1]
```diff
@@ -43,8 +46,9 @@ predicate mayAddNullTerminator(Expr e, VariableAccess va) {
   )
   or
   // Assignment to another stack variable
-  exists(Expr e0 |
-    mayAddNullTerminatorHelper(pragma[only_bind_into](e), va, pragma[only_bind_into](e0)) and
+  exists(StackVariable v0, Expr e0 |
+    mayAddNullTerminatorHelper(e, va, v0) and
+    mayAddNullTerminator(pragma[only_bind_into](e0), pragma[only_bind_into](v0.getAnAccess())) and
```
Comment (MathiasVP) [inline cpp/ql/lib/semmle/code/cpp/commons/NullTermination.qll:51]: Ah, I see why this speeds it up. Previously, we were computing an SCC consisting of `{mayAddNullTerminator, mayAddNullTerminatorHelper}`, but now we're computing `mayAddNullTerminatorHelper` first, and then evaluating an SCC _just_ containing `mayAddNullTerminator`. So this doesn't change any query results. Is that a correct description of what's going on?

File: cpp/ql/lib/semmle/code/cpp/commons/NullTermination.qll [hunk 2]
```diff
@@ -3,11 +3,14 @@ private import semmle.code.cpp.models.interfaces.ArrayFunction
...
 import semmle.code.cpp.dataflow.DataFlow
 
-private predicate mayAddNullTerminatorHelper(Expr e, VariableAccess va, Expr e0) {
-  exists(StackVariable v0, Expr val |
-    exprDefinition(v0, e, val) and
-    val.getAChild*() = va and
-    mayAddNullTerminator(e0, v0.getAnAccess())
+/**
+ * Holds if the expression `e` assigns something including `va` to a
+ * stack variable `v0` that is later null terminated at `e0`.
```
Comment (MathiasVP) [inline cpp/ql/lib/semmle/code/cpp/commons/NullTermination.qll:8]:
```suggestion
 * Holds if the expression `e` assigns something including `va` to a
 * stack variable `v0`.
```
`e0` isn't a parameter anymore, and I guess the "later null terminated"-part is now moved to `mayAddNullTerminator`, right?

**Review by MathiasVP — COMMENTED** [review]
LGTM (with the exception of a small QLDoc nitpick).

File: cpp/ql/lib/semmle/code/cpp/commons/NullTermination.qll (hunk 2 above)
Comment (geoffw0) [reply cpp/ql/lib/semmle/code/cpp/commons/NullTermination.qll:8]: Good point, well spotted.

File: cpp/ql/lib/semmle/code/cpp/commons/NullTermination.qll (hunk 1 above)
Comment (geoffw0) [reply cpp/ql/lib/semmle/code/cpp/commons/NullTermination.qll:51]: Yes, that's my understanding as well.  The tricky part is that we may be computing a bigger (less constrained)  `mayAddNullTerminatorHelper` relation in the first place, but I haven't seen that be a problem.

**Review by MathiasVP — APPROVED** [review]
LGTM!

**Pull request merged**